libclamav_rust/src/alz.rs

```diff
@@ -0,0 +1,551 @@
+    File = 0x20,
+}
+
+impl AlzLocalFileHeader {
+    const fn is_encrypted(&self) -> bool {
+        0 != (self.head.file_descriptor & 0x1)
+    }
+
+    const fn is_data_descriptor(&self) -> bool {
+        0 != (self.head.file_descriptor & 0x8)
+    }
+
+    const fn is_directory(&self) -> bool {
+        0 != ((AlzFileAttribute::Directory as u8) & self.head.file_attribute)
+    }
+
+    const fn _is_file(&self) -> bool {
+        0 != ((AlzFileAttribute::File as u8) & self.head.file_attribute)
+    }
+
+    const fn _is_readonly(&self) -> bool {
+        0 != ((AlzFileAttribute::Readonly as u8) & self.head.file_attribute)
+    }
+
+    const fn _is_hidden(&self) -> bool {
+        0 != ((AlzFileAttribute::Hidden as u8) & self.head.file_attribute)
+    }
+
+    fn _dump(&self) {
+        println!(
+            "self.start_of_compressed_data = {}",
+            self.start_of_compressed_data
+        );
+
+        println!(
+            "self.head.file_name_length = {:x}",
+            self.head.file_name_length
+        );
+        println!(
+            "self.head.file_attribute = {:02x}",
+            self.head.file_attribute
+        );
+        println!("self.head.file_time_date = {:x}", self.head.file_time_date);
+        println!(
+            "self.head.file_descriptor = {:x}",
+            self.head.file_descriptor
+        );
+        println!("self.head.unknown = {:x}", self.head.unknown);
+
+        println!("self.compression_method = {:x}", self.compression_method);
+        println!("self.unknown = {:x}", self.unknown);
+        println!("self.file_crc = {:x}", self.file_crc);
+        println!("self.compressed_size = {:x}", self.compressed_size);
+        println!("self.uncompressed_size = {:x}", self.uncompressed_size);
+
+        println!("self.file_name = {}", self.file_name);
+
+        print!("self.enc_chk = ");
+        for i in 0..ALZ_ENCR_HEADER_LEN {
+            if 0 != i {
+                print!(" ");
+            }
+            print!("{}", self.enc_chk[i as usize]);
+        }
+        println!();
+
+        println!("is_encrypted = {}", self.is_encrypted());
+        println!("is_data_descriptor = {}", self.is_data_descriptor());
+
+        println!();
+    }
+
+    pub const fn new() -> Self {
+        Self {
+            head: AlzLocalFileHeaderHead {
+                file_name_length: 0,
+                file_attribute: 0,
+                file_time_date: 0,
+                file_descriptor: 0,
+                unknown: 0,
+            },
+
+            compression_method: 0,
+            unknown: 0,
+            file_crc: 0,
+            compressed_size: 0,
+            uncompressed_size: 0,
+            file_name: String::new(),
+            enc_chk: [0; ALZ_ENCR_HEADER_LEN as usize],
+            start_of_compressed_data: 0,
+        }
+    }
+
+    pub fn parse_internal(
+        &mut self,
+        cursor: &mut std::io::Cursor<&Vec<u8>>,
+    ) -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
This is pretty hardcore, using `Box<dyn ... >`. But there is an easier way to pass back different types of errors.